spring-kafka/src/main/java/org/springframework/kafka/requestreply/AggregatingReplyingKafkaTemplate.java

```diff
@@ -0,0 +1,247 @@
+		List<ConsumerRecord<K, Collection<ConsumerRecord<K, R>>>> completed = new ArrayList<>();
+		data.forEach(record -> {
+			Header correlation = record.headers().lastHeader(KafkaHeaders.CORRELATION_ID);
+			if (correlation == null) {
+				this.logger.error("No correlationId found in reply: " + record
+						+ " - to use request/reply semantics, the responding server must return the correlation id "
+						+ " in the '" + KafkaHeaders.CORRELATION_ID + "' header");
+			}
+			else {
+				CorrelationKey correlationId = new CorrelationKey(correlation.value());
+				synchronized (this) {
+					if (isPending(correlationId)) {
+						List<ConsumerRecord<K, R>> list = addToCollection(record, correlationId).stream()
+								.map(entry -> entry.getRecord())
+								.collect(Collectors.toList());
+						if (this.releaseStrategy.test(list)) {
+							ConsumerRecord<K, Collection<ConsumerRecord<K, R>>> done =
+									new ConsumerRecord<>(AGGREGATED_RESULTS_TOPIC, 0, 0L, null, list);
+							done.headers()
+								.add(new RecordHeader(KafkaHeaders.CORRELATION_ID, correlationId.getCorrelationId()));
+							this.pending.remove(correlationId);
+							checkOffsetsAndCommitIfNecessary(list, consumer);
+							completed.add(done);
+						}
+					}
+					else {
+						logLateArrival(record, correlationId);
+					}
+				}
+			}
+		});
+		if (completed.size() > 0) {
+			super.onMessage(completed);
+		}
+	}
+
+	@Override
+	protected synchronized boolean handleTimeout(CorrelationKey correlationId,
+			RequestReplyFuture<K, V, Collection<ConsumerRecord<K, R>>> future) {
+
+		Set<RecordHolder<K, R>> removed = this.pending.remove(correlationId);
+		if (removed != null && this.returnPartialOnTimeout) {
+			List<ConsumerRecord<K, R>> list = removed.stream()
+				.map(entry -> entry.getRecord())
+				.collect(Collectors.toList());
+			future.set(new ConsumerRecord<>(PARTIAL_RESULTS_AFTER_TIMEOUT_TOPIC, 0, 0L, null, list));
+			return true;
+		}
+		else {
+			return false;
+		}
+	}
+
+	private void checkOffsetsAndCommitIfNecessary(List<ConsumerRecord<K, R>> list,
+			Consumer<?, ?> consumer) {
+
+		list.forEach(record -> this.offsets.compute(
+			new TopicPartition(record.topic(), record.partition()),
+				(k, v) -> v == null ? record.offset() + 1 : Math.max(v, record.offset() + 1)));
+		if (this.pending.isEmpty() && !this.offsets.isEmpty()) {
+			consumer.commitSync(this.offsets.entrySet().stream()
+					.collect(Collectors.toMap(
+											Map.Entry::getKey,
+											entry -> new OffsetAndMetadata(entry.getValue()))),
+								this.commitTimeout);
+			this.offsets.clear();
+		}
+	}
+
+	@SuppressWarnings({ "rawtypes", "unchecked" })
+	private Set<RecordHolder<K, R>> addToCollection(ConsumerRecord record, CorrelationKey correlationId) {
+		Set<RecordHolder<K, R>> set = this.pending.computeIfAbsent(correlationId, id -> new LinkedHashSet<>());
+		set.add(new RecordHolder<>(record));
```

Comment:
Looks like we don't need `RecordHolder` any more...

Reply:
We need `RecordHolder` because we avoid duplicates via a `Set`. `ConsumerRecord` does not implement `hashCode()` and `equals()`. `RecordHolder` `hashCode()` and `equals()` are based on topic/partition/offset.